### PR TITLE
slate: Add approximate flop counts to kernels

### PR DESCRIPTION
**Pull request by wence-**
A slate expression has a flop count consisting of the flops to
assemble the terminal data (approximately determined from the TSFC
kernels), and the flops to compute the expression on assembled data.
This latter is just matrix algebra, so easy to do.

@rckirby use this to get flops counts for slate kernels doing static condensation.

### CODE REVIEW

**Comment by rckirby** [comment]
Thanks, @wence- 
Are we sure we're counting facet flops right?  You multiply an interior facet integral by the number of facets a reference cell has.  I think that double-counts in some sense because those facets are shared by two cells?

**Comment by wence-** [comment]
In particular, interior facet integrals have been rewritten into exterior facet integrals before being fed to tsfc: they're exterior to the domain of the cell.

**Comment by rckirby** [comment]
Ok, maybe I misunderstood -- are these computed by visiting facets of cells
rather than in a separate loop over the facets?

**Comment by thomasgibson** [comment]
Correct. Everything is from the perspective of the cell. At the moment, Slate kernels know nothing about the adjacent cell. It can, however, distinguish between a facet that's interior or on the domain boundary (including sub-domains). However, all facet integrals are treated as "exterior to the cell."

File: firedrake/slate/slac/kernel_builder.py [hunk 1]
```diff
@@ -128,6 +130,84 @@ def dimension(e):
...
+        nfacets = self.expression.ufl_domain().ufl_cell().num_facets()
+        for ctx in self.context_kernels:
+            itype = ctx.original_integral_type
+            for k in ctx.tsfc_kernels:
+                kinfo = k.kinfo
+                if itype == "cell":
+                    flops += kinfo.kernel.num_flops
+                elif itype.startswith("interior_facet"):
+                    # Executed once per facet (approximation)
+                    flops += kinfo.kernel.num_flops * nfacets
```
Comment (thomasgibson) [inline firedrake/slate/slac/kernel_builder.py]: Just noting here that this is a bit tricky if the integral is defined over facet sub-domains. In which case, not all facets will be included in the local computation.

Comment (wence-) [reply firedrake/slate/slac/kernel_builder.py]: Yes, there's no real way to get this exactly right, short of runtime counting in some way.

**Pull request merged**